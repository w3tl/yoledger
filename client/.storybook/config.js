import React from 'react';
import { configure, addDecorator } from '@storybook/react';
import 'semantic-ui-css/semantic.min.css';
import { MemoryRouter } from 'react-router-dom';
import apolloDecorator from './apolloDecorator';

addDecorator(apolloDecorator);
addDecorator(story => (<MemoryRouter initialEntries={['/']}>{story()}</MemoryRouter>));

const req = require.context('../stories', true, /\.jsx$/);

function loadStories() {
  req.keys().forEach((filename) => req(filename))
}

configure(loadStories, module);
