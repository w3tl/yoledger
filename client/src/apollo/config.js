// eslint-disable-next-line no-undef
const hostname = window && window.location && window.location.hostname;
// eslint-disable-next-line import/prefer-default-export
export const GRAPHQL_ENDPOINT = process.env.REACT_APP_GRAPHQL_ENDPOINT || `https://api.${hostname}/graphql`;
