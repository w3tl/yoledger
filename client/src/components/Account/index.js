import AccountInput from './AccountInput';

export {
  AccountInput,
};
