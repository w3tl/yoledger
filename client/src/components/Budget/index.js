import BudgetPage from './BudgetPage';
import propTypes from './propTypes';

export {
  BudgetPage,
  propTypes,
};
