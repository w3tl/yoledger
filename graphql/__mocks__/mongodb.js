import {
  users,
  transactions as transactionsJson,
  accounts,
  budgets as budgetsJson,
} from '../mocks';

const mongodb = jest.genMockFromModule('mongodb');

const addCreatedAt = ({ createdAt, ...other }) => ({ ...other, createdAt: new Date(createdAt) });
const addDate = ({ date, ...other }) => ({ ...other, date: new Date(date) });
const transactions = transactionsJson.map(addCreatedAt);
const budgets = budgetsJson.map(addDate);

const mockSession = {
  startTransaction: () => Promise.resolve(),
  commitTransaction: () => Promise.resolve(),
  abortTransaction: () => Promise.reject({ // eslint-disable-line prefer-promise-reject-errors
    errorLabels: ['TransientTransactionError'],
  }),
};

const mockCollections = {
  users: {
    findOne: ({ _id }) => Promise.resolve(users.find(u => u._id === _id)),
  },
  transactions: {
    find: () => ({
      skip: skip => ({
        limit: limit => ({
          toArray: () => transactions.slice(skip, skip + limit),
        }),
      }),
    }),
    findOne: ({ _id }) => Promise.resolve(transactions.find(t => t._id === _id)),
    insertOne: (transaction) => {
      transactions.push({ _id: '50', ...transaction });
      return Promise.resolve({ insertedId: '50' });
    },
    findOneAndDelete: ({ id }) => Promise.resolve({ value: transactions.find(t => t.id === id) }),
    updateOne: () => Promise.resolve(),
  },
  accounts: {
    find: ({ type }) => ({
      toArray: () => Promise.resolve(accounts.filter(acc => type === acc.type)),
    }),
    findOne: ({ name, _id }) => Promise.resolve(
      accounts.find(acc => name === acc.name || _id === acc._id),
    ),
    insertOne: (account) => {
      accounts.push({ _id: '44', ...account });
      return Promise.resolve({ insertedId: '44' });
    },
    updateOne: () => Promise.resolve(),
  },
  budget: {
    findOne: ({ _id, date }) => {
      if (_id) { return Promise.resolve(budgets.find(t => t._id === _id)); }
      if (date) { return Promise.resolve(budgets.find(t => t.date.getTime() === date.getTime())); }
      return budgets[0];
    },
    updateOne: ({ date, account }, { $set }) => {
      const idx = budgets.find(b => b.date.getTime() === date.getTime() && b.account === account);
      if (idx) {
        budgets[idx] = { ...budgets[idx], ...$set.modificator };
      }
      return Promise.resolve({ upsertedId: budgets[idx]._id });
    },
    aggregate: () => ({
      toArray: () => ([{
        _id: { account: 'Train' },
        allocations: [
          { date: new Date('2018-06-10'), amount: 14, balance: 0 },
        ],
      }, {
        _id: { account: 'Food' },
        allocations: [
          { date: new Date('2018-05-10'), amount: 50, balance: 0 },
          { date: new Date('2018-05-25'), amount: 55, balance: 0 },
          { date: new Date('2018-06-10'), amount: 60, balance: 0 },
        ],
      }]),
    }),
  },
};


mongodb.connection = {
  db: () => ({
    collection: name => mockCollections[name],
  }),
  withSession: callback => callback(mockSession),
};

module.exports = mongodb;
