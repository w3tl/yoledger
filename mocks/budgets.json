[
  {
    "date": "2018-05-25",
    "account": "Food",
    "amount": 9,
    "balance": 0
  },
  {
    "date": "2018-05-25",
    "account": "Train",
    "amount": 29,
    "balance": 20
  },
  {
    "date": "2018-06-10",
    "account": "Food",
    "amount": 10,
    "balance": 0
  },
  {
    "date": "2018-06-10",
    "account": "Train",
    "amount": 30,
    "balance": 20
  },
  {
    "date": "2018-06-25",
    "account": "Food",
    "amount": 11,
    "balance": 1
  }
]
