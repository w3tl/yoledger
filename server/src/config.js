const convict = require('convict');

const config = convict({
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test', 'stage'],
    default: 'development',
    env: 'NODE_ENV',
  },
  port: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3030,
    env: 'PORT',
    arg: 'port',
  },
  credentials: {
    algorithm: {
      doc: 'Current crypt algorithm',
      env: 'CRED_ALG',
      default: 'default',
    },
  },
  mongoUri: {
    doc: 'MongoDB connection string',
    format: '*',
    default: 'mongodb://localhost',
    env: 'MONGO_URI',
  },
});

// Load environment dependent configuration
const env = config.get('env');
config.loadFile(`config/${env}.json`);
// Perform validation
config.validate({ allowed: 'strict' });

module.exports = config;
