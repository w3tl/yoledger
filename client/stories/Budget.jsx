/* eslint-disable react/prop-types */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Table } from 'semantic-ui-react';
import { BudgetPage } from '../src/components/Budget';
import RawBudgetTable from '../src/components/Budget/BudgetTable';
import RawBudgetTableHeader from '../src/components/Budget/BudgetTableHeader';
import RawBudgetTableFooter from '../src/components/Budget/BudgetTableFooter';
import RawBudgetTableRow from '../src/components/Budget/BudgetTableRow';
import { accounts, budgets } from '../src/components/Budget/mockData';
import { getPeriods } from '../src/components/Budget/utils';

const periods = getPeriods('2018-06-10', 6, {
  type: 'dayOfMonth', options: [10, 25],
});

storiesOf('Budget/Page', module)
  .add('with budget', () => <BudgetPage initialState={{ periods }} />);

const Row = ({ periods: cols, account }) => (
  <Table.Row>
    <Table.Cell>{account.name}</Table.Cell>
    {cols.map(period => <Table.Cell key={period}>Cell</Table.Cell>)}
  </Table.Row>
);

const BudgetTable = props => (
  <RawBudgetTable periods={periods} accounts={accounts} {...props} />
);

storiesOf('Budget/Table', module)
  .add('default', () => <BudgetTable row={Row} />)
  .add('without accounts', () => <BudgetTable row={Row} accounts={[]} />)
  .add('loading', () => <BudgetTable row={Row} loading />);

const BudgetTableHeader = props => (
  <RawBudgetTableHeader periods={periods} onPeriodChange={action('onPeriodChange')} {...props} />
);

storiesOf('Budget/Table/Header', module)
  .addDecorator(story => (
    <Table fixed columns={periods.length + 1}>
      {story()}
    </Table>
  ))
  .add('default', () => <BudgetTableHeader />);

const BudgetTableFooter = props => (
  <RawBudgetTableFooter
    accounts={accounts}
    length={periods.length}
    onCreate={action('onCreate')}
    {...props}
  />
);

storiesOf('Budget/Table/Footer', module)
  .addDecorator(story => (
    <Table fixed columns={periods.length + 1}>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>{accounts[0].name}</Table.HeaderCell>
          <Table.HeaderCell>100</Table.HeaderCell>
          <Table.HeaderCell colSpan={periods.length - 1} />
        </Table.Row>
      </Table.Header>
      {story()}
    </Table>
  ))
  .add('default', () => <BudgetTableFooter />)
  .add('when click add', () => <BudgetTableFooter initialState={{ isAddAccount: true }} />)
  .add('when click add and enter new account', () => <BudgetTableFooter initialState={{ isAddAccount: true, newAccount: 'New account' }} />)
  .add('when click add and enter existing account', () => <BudgetTableFooter initialState={{ isAddAccount: true, newAccount: accounts[0].name }} />)
  .add('loading', () => <BudgetTableFooter loading />);

const BudgetTableRow = props => (
  <RawBudgetTableRow
    account={accounts[0]}
    periods={periods}
    budget={budgets.filter(b => b.account.name === accounts[0].name)}
    onUpdate={action('onUpdate')}
    {...props}
  />
);

storiesOf('Budget/Table/Row', module)
  .addDecorator(story => (
    <Table fixed columns={periods.length + 1}>
      <Table.Body>{story()}</Table.Body>
    </Table>
  ))
  .add('default', () => <BudgetTableRow />)
  .add('loading', () => <BudgetTableRow loading />);
