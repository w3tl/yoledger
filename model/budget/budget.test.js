import { MongoClient, ObjectId } from 'mongodb';
import config from '../../config';
import Budget from './index';
import Account from '../accounts';
import budgets from '../../mocks/budgets';
import accounts from '../../mocks/accounts';

let db;
let connection;
let budgetModel;
let budgetId;
let accountModel;

beforeAll(async () => {
  connection = await MongoClient.connect(config.get('mongoUri'), {
    connectWithNoPrimary: true,
    useNewUrlParser: true,
  });
  db = connection.db('dbbudgets');
  budgetModel = new Budget(db, 'user1');
  accountModel = new Account(db, 'user1');
  await accountModel.init();
  const expenseAccount = accounts.find(({ name }) => name === budgets[0].account);
  await accountModel.create(expenseAccount);
  await budgetModel.init();
  budgetId = await budgetModel.addBudget(budgets[0]);
});

afterAll(async () => {
  await connection.close();
});

describe('budget', () => {
  test('checkAccount', async () => {
    const budgetDoc = await budgetModel.findOne({ _id: budgetId });

    expect.assertions(2);
    await expect(budgetModel.checkAccount('blabla'))
      .rejects
      .toThrowError('Account not found');

    let error;
    try {
      await budgetModel.checkAccount(budgetDoc.account);
    } catch (err) {
      error = err;
    }
    expect(error).toBeUndefined();
  });

  test('Must add period', async () => {
    const budgetDoc = await budgetModel.findOne({ _id: budgetId });
    expect(budgetDoc).toMatchSnapshot({ _id: expect.any(ObjectId) }, 'New budget');
  });

  test('Must update period', async () => {
    const newAccount = accounts.find(({ name }) => name !== budgets[0].account);
    await accountModel.create(newAccount);
    const { result: { nModified } } = await budgetModel.updateBudget(budgetId, {
      account: newAccount.name,
      date: new Date('2018-01-10'),
      amount: 200,
    });
    expect(nModified).toEqual(1);
    const budgetDoc = await budgetModel.findOne({ _id: budgetId });
    expect(budgetDoc).toMatchSnapshot({ _id: expect.any(ObjectId) }, 'Updated budget period');
  });
});
