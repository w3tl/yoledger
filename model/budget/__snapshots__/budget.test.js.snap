// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget periods func should return correct array: 2018-01-9 1`] = `
Array [
  2018-01-10T00:00:00.000Z,
  2018-01-25T00:00:00.000Z,
  2018-02-10T00:00:00.000Z,
]
`;

exports[`budget periods func should return correct array: 2018-01-11 1`] = `
Array [
  2018-01-25T00:00:00.000Z,
  2018-02-10T00:00:00.000Z,
  2018-02-25T00:00:00.000Z,
]
`;

exports[`budget periods func should return correct array: 2018-01-26 1`] = `
Array [
  2018-02-10T00:00:00.000Z,
  2018-02-25T00:00:00.000Z,
  2018-03-10T00:00:00.000Z,
]
`;

exports[`budget upsertBudget should add a budget to the existing period: should contain "twix" account with amount of 15 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": "Food",
  "amount": 14,
  "balance": 0,
  "date": 2018-05-25T00:00:00.000Z,
  "userId": "admin",
}
`;

exports[`budget upsertBudget should create a non-existent period with a budget: should contain "twix" account with amount of 15 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": "twix",
  "amount": 15,
  "date": 2018-01-13T00:00:00.000Z,
  "userId": "admin",
}
`;

exports[`budget upsertBudget should update the amount in the existing period: should contain "Food" account with amount of 14 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": "Food",
  "amount": 14,
  "balance": 0,
  "date": 2018-05-25T00:00:00.000Z,
  "userId": "admin",
}
`;
